### DISCUSSION TITLE
How can get statistic from Smart Plug?

### DISCUSSION THREAD

**Discussion by EugeneRymarev** [jasonacox/tinytuya#107] [Q&A] [answered]
Hello everyone!
I have many Smart Plugs.
I want to collect statistics from them.
How can this be implemented?
Where can I find an example to work with them?
Preferably via API, but LAN is also possible.

Do I need to connect my Smart Plugs to a Developer Account?

**Comment by jasonacox** [comment, marked as answer]
Hi @EugeneRymarev -

* LAN access: To get the statistics (status) of those Tuya smartplugs over LAN you will need the device LocalKey. 
* Cloud API access: To get statistics via the Tuya Cloud you will need to set up an IoT account.  

For both, that requires that you register them with the Smart Life App or Tuya Smart App and to set up an IoT account with Tuya..  See the setup instructions: https://github.com/jasonacox/tinytuya#tuya-device-preparation

LAN 
```python
    import tinytuya

    d = tinytuya.OutletDevice('DEVICE_ID_HERE', 'IP_ADDRESS_HERE', 'LOCAL_KEY_HERE')
    d.set_version(3.3)
    data = d.status() 
    print('set_status() result %r' % data)
```

CloudAPI
```python
import tinytuya

# Connect to Tuya Cloud
# c = tinytuya.Cloud()  # uses tinytuya.json 
c = tinytuya.Cloud(
        apiRegion="us", 
        apiKey="xxxxxxxxxxxxxxxxxxxx", 
        apiSecret="xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", 
        apiDeviceID="xxxxxxxxxxxxxxxxxxID")

# Display list of devices
devices = c.getdevices()
print("Device List: %r" % devices)

# Select a Device ID to Test
id = "xxxxxxxxxxxxxxxxxxID"

# Display Status of Device
result = c.getstatus(id)
print("Status of device:\n", result)
```

**Reply by EugeneRymarev** [reply]
I know you are not manager of Tuya IoT, but i can't download IoT Device Manager APK. Can you upload it here or give me link?
Link from QR code Access Denied

**Comment by jasonacox** [comment]
Hi @EugeneRymarev - You should not need the Device Manager APK.  You should be able to connect your devices and link your IoT account with the "Smart Life" App - Here: https://play.google.com/store/apps/details?id=com.tuya.smartlife

You will use that app to scan the QR code on the iot.tuya.com website per [STEP 3 on the instructions](https://github.com/jasonacox/tinytuya#setup-wizard---getting-local-keys) which will connect your devices to the IoT account you create.

**Reply by EugeneRymarev** [reply]
Nope, Smart Life app not linked. Just open url. IoT Device Manager linked.

**Reply by EugeneRymarev** [reply]
Wrong screenshot in instruction.
https://user-images.githubusercontent.com/836718/138599218-e2202d7b-d9a0-4b18-bf40-c01567944da3.png
not selected needed tab.

**Reply by EugeneRymarev** [reply]
Yes, Smart Life application is now linked.
But how can i link digma smart life - i use this app, because at first setup i install it.
Or maybe you know easy way to transfer devices from digma app to smart life app?
This apps almost same just another design, but works with same protocols.

When i scan QR with Digma App i always receive in Tuya "QR code has expired" in web.

**Reply by EugeneRymarev** [reply]
Nevermind. For now i can't use Tuya API because it's not free.
I apologize for the disturbance.

**Reply by jasonacox** [reply]
Hi @EugeneRymarev  - what is not free?  I haven't paid for anything to use the Tuya API.  Are you looking at their commercial offerings instead of the free trial option?

Also, on the screenshot. Thanks for that feedback.  What this the process you tried? 

![https://user-images.githubusercontent.com/836718/138599473-066bb1b7-99b7-4540-983a-a0397b199448.png](https://user-images.githubusercontent.com/836718/138599473-066bb1b7-99b7-4540-983a-a0397b199448.png)

**Reply by EugeneRymarev** [reply]
![image](https://user-images.githubusercontent.com/6260309/155654129-4581f607-b268-4061-ac1b-80d1491db847.png)
This QR code is right for Smart Life App.

**Reply by EugeneRymarev** [reply]
`c.getdevices().devices` gives me this error
```
Device List: {'Error': 'Error Response from Tuya Cloud', 'Err': '913', 'Payload': 'Unable to get device list'}
```
`c.getstatus(id_)` gives me this error
```
Status of device:
 {'code': 28841101, 'msg': 'No permissions. This API is not subscribed.', 'success': False, 't': 1645763611021, 'tid': '15ad2e8195f411ecb018e66f3a529df6'}
```
To subscribe need to pay.

**Comment by jasonacox** [comment]
Thanks for the information. When you set it up, did you select the free API option?

<img width="247" alt="Screen Shot 2022-02-24 at 8 46 11 PM" src="https://user-images.githubusercontent.com/836718/155658596-502a4a0e-99e4-4f6c-a2f4-e8576610dc0d.png">

I created a PDF of the screenshots to set up... is this similar to what you did?  

[Tuya IoT API Setup.pdf](https://github.com/jasonacox/tinytuya/files/8138703/Tuya.IoT.API.Setup.pdf)

**Reply by EugeneRymarev** [reply]
Everything is same except:
1. I choose Individual developer instead skip this step
2. I scan another QR code to pair accounts - i show it [previously](https://github.com/jasonacox/tinytuya/discussions/107#discussioncomment-2248976)

**Reply by EugeneRymarev** [reply]
I create project again and now all works correctly. Thanks.

**Reply by jasonacox** [reply]
That's great news!  Did you follow the PDF steps I sent?  If so, should I add a link to the README setup instruction to those slides?

**Reply by EugeneRymarev** [reply]
Yes, but that same steps, like i do earlier.

Another question:
![image](https://user-images.githubusercontent.com/6260309/155668749-bde72f7a-c475-4818-94b8-d34df18c7772.png)
How can i receive statistics from smart plug? Not see this in _Cloud_ class

**Reply by jasonacox** [reply]
Thank you!  

I updated this slide:
<img width="732" alt="image" src="https://user-images.githubusercontent.com/836718/155827671-44d5fce4-0119-4d0e-a224-ef3715fafc24.png">

And the PDF:
[Tuya IoT API Setup.pdf](https://github.com/jasonacox/tinytuya/files/8145832/Tuya.IoT.API.Setup.pdf)


The power metrics should show up in the status() or getdps() responses:

```python
# Display Status of Device
result = c.getstatus(id)
print("Status of device: %r\n" % result)

# Display Data Points (DPS)
result = c.getdps(id)
print("DPS: %r\n" % result)

```

**Reply by EugeneRymarev** [reply]
> The power metrics should show up in the status() or getdps() responses

[That's all](https://pastebin.com/6YdmVe5S) what i receive. This information does not contain the statistics I need.

**Comment by jasonacox** [comment]
I'm sorry, I misunderstood.  If you are looking for current state, voltage (V), current (amps) and power (W), that data you got from getstatus(id) should work. 

If you are wanting to access historical power data (kWh), I'm afraid I haven't figured that out.  You can look at https://iot.tuya.com/cloud/explorer to see if you can find the right API call that would have that payload.  Let me know if you find it and I can add that to TinyTuya.

**Reply by peterodro** [reply]
i can't get "Electric today (kWh)" from cloud and also not locally.
Add_ele (DPS 17) is something else. I suppose it is energy consumed between 2 updates. not sure.
No any way to get "Electric today (kWh)" as see on APK?
